src/mesh.cpp

```diff
@@ -1852,6 +1963,225 @@ UnstructuredMesh::write(std::string base_filename) const {
+{
+  filename_ = filename;
+  initialize();
+}
+
+void LibMesh::initialize()
+{
+  if (!settings::LMI) {
+    fatal_error("Attempting to use an unstructured mesh without a libMesh::LibMeshInit object.");
+  }
+
+  m_ = std::make_unique<libMesh::Mesh>(settings::LMI->comm(), n_dimension_);
+  m_->read(filename_);
+  m_->prepare_for_use();
+
+  // ensure that the loaded mesh is 3 dimensional
+  if(m_->mesh_dimension() != n_dimension_) {
+    fatal_error(fmt::format("Mesh file {} specified for use in an unstructured mesh is not a 3D mesh.", filename_));
+  }
+
+  // create an equation system for storing values
+  eq_system_name_ = "mesh_" + std::to_string(id_) + "_system";
+
+  equation_systems_ = std::make_unique<libMesh::EquationSystems>(*m_);
+  libMesh::ExplicitSystem& eq_sys =
+    equation_systems_->add_system<libMesh::ExplicitSystem>(eq_system_name_);
+
+
+  #ifdef _OPENMP
+  int n_threads = omp_get_max_threads();
+  #else
+  int n_threads = 1;
+  #endif
+
+  for (int i = 0; i < n_threads; i++) {
+      PL_.emplace_back(m_->sub_point_locator());
+      PL_.back()->set_contains_point_tol(FP_COINCIDENT);
+      PL_.back()->enable_out_of_mesh_mode();
+  }
+
+  // store first element in the mesh to use as an offset for bin indices
+  first_element_ = *m_->elements_begin();
+
+  // bounding box for the mesh for quick rejection checks
+  bbox_ = libMesh::MeshTools::create_bounding_box(*m_);
+}
+
+Position
+LibMesh::centroid(int bin) const
+{
+  auto& elem = this->get_element_from_bin(bin);
+  auto centroid = elem.centroid();
+  return {centroid(0), centroid(1), centroid(2)};
+}
+
+std::string LibMesh::library() const { return "libmesh"; }
+
+int LibMesh::n_bins() const
+{
+  return m_->n_elem();
+}
+
+int LibMesh::n_surface_bins() const
+{
+  int n_bins = 0;
+  for (int i = 0; i < this->n_bins(); i++) {
+    const libMesh::Elem& e = get_element_from_bin(i);
+    n_bins += e.n_faces();
+    // if this is a boundary element, sure to count boundary faces
+    // twice
+    for (int j = 0; j < e.n_sides(); j++) {
+      if (e.neighbor_ptr(j) == NULL) { n_bins++; }
+    }
+  }
+  return n_bins;
+}
+
+void
+LibMesh::add_score(const std::string& var_name)
+{
+  // check if this is a new variable
+  std::string value_name = var_name + "_mean";
+  if (!variable_map_.count(value_name)) {
+    auto& eqn_sys = equation_systems_->get_system(eq_system_name_);
+    auto var_num = eqn_sys.add_variable(value_name, libMesh::CONSTANT, libMesh::MONOMIAL);
+    variable_map_[value_name] = var_num;
+  }
+
+  std::string std_dev_name = var_name + "_std_dev";
+  // check if this is a new variable
+  if (!variable_map_.count(std_dev_name)) {
+    auto& eqn_sys = equation_systems_->get_system(eq_system_name_);
+    auto var_num = eqn_sys.add_variable(std_dev_name, libMesh::CONSTANT, libMesh::MONOMIAL);
+    variable_map_[std_dev_name] = var_num;
+  }
+}
+
+void
+LibMesh::remove_score(const std::string& var_name)
+{
+  return;
+}
```

Comment:
Yep. I need to rethink how the mesh writing is done. It's very different between MOAB and libMesh, so I think the best thing to do might be to contain all of the writing in a single method instead of the many methods that are used now.